Zabuza (Francesco Pannullo) - obj ctor completed

novaryth opened the pull request.

novaryth committed:
Update obj ctor

novaryth committed:
Update obj parser

Added ctor obj triangles

novaryth committed:
Minor update

novaryth committed:
Minor update

fdefelici reviewed (changes requested):
Implementation seems fine. 

Possibly for making obj_parser a real single-header library, try to add an OBJ_IMPLEMENTATION macro to separate declarations from definitions (like stb_image does or as we saw in clove-unit)

About naming:
`obj_parser` is more a entity then an action. If you want to change the "namespace" part ok, so eventually it should be something like `obj_parser_parse` or `obj_parser_read`. Even the variable `obj_parse` should be the result of the parse action so the `obj`

Futhermore you could make few codes improvements:
- try remove duplication for the first strtok call related to v, vt, vn and f
- replace `rewind` with more secure `fseek` call
- replace `atoi` and `atof` with more secure `strtol and `strtof`

novaryth committed:
Requested changes

- added OBJ_IMPLEMENTATION macro
- changed obj_parser namespace
- removed duplication for the first strtok call related to v, vt, vn and f
-replaced rewind with more secure fseek call
-replaced atoi and atof with more secure strtol and strtof`

novaryth committed:
Minor update

- Fix naming vars
- Added param desc to function

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -86,87 +108,65 @@ obj_t *obj_parse(const char *file_name)
-
-            token = strtok_s(NULL, " ", &remaining_tokens);
-            obj->v[v_index].z = atof(token);
+            obj->v[v_index].x = strtof(strtok_s(buffer, "v ", &remaining_tokens), NULL);
```
smart using strtof side effect which read until it finds numbers, but could be better favor straitgh forwared code, where you try to just parse the only part that is a number.

So here the suggestion  (even to remove definitly the strtok_s duplication in all scenarios ( `f` included) coud be:
- move char *remaining_tokens and a strtok to split on " " (space) to take the first token (that will be `v`, `vt`, `vn` and `f`)  just after the while (before the first if)
- change the string compare to compare just to v, vt, etc.... (now you just need `strcmp`)
- then within each if you have just to take care of the token that contains values

Bonus for performance:
- use 1 `if ` and 3 `else if` instead of 4 `if` (otherwise all of 4 are always evaluated)

novaryth committed:
Update obj ctor

Added the requested optimizations

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -86,87 +108,64 @@ obj_t *obj_parse(const char *file_name)
-                token = strtok_s(NULL, " ", &remaining_tokens);
-                obj->v_info[f_index + i].vn_index = atoi(token);
+                // Get values from buffer and fill struct
+                sscanf_s(remaining_tokens, "%d/%d/%d",
```
In general, I wouldn't mix strtok and scanf. For sure it works, but just  having one common way in the code to parse tokens in this scenario could be preferred my opionion.

In particular on this last scenario when parsing faces, it leads to a less readable solution where mixing for/while and pointer aritmetic.

novaryth replied:
> In general, I wouldn't mix strtok and scanf. For sure it works, but just having one common way in the code to parse tokens in this scenario could be preferred my opionion.

I could eliminate the internal while loop with a memmove to eliminate the already assigned values each time
What do you think?

> In particular on this last scenario when parsing faces, it leads to a less readable solution where mixing for/while and pointer aritmetic.

Regarding the use of scanf I thought it would be useful. I'll try to find another way
Thanks for feedbacks ;)

fdefelici replied:
It is good you try to find other solutions and compare them to look the more efficient.

For this case I think that just strtok can do the job, keeping the code easy to read.

novaryth replied:
I restored the use of stroke, everything should be ok now.

novaryth committed:
Update obj ctor

Reimplement strok

zabuza-momochi commented:
I restored the use of stroke, everything should be ok now.

fdefelici committed:
add counters test for stoormtroop
